File: analysis/csrc/domain/services/device_context/device_info.cpp
/* -------------------------------------------------------------------------
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is part of the MindStudio project.
 *
 * MindStudio is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *
 *    http://license.coscl.org.cn/MulanPSL2
 *
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * -------------------------------------------------------------------------*/
#include "nlohmann/json.hpp"
#include "analysis/csrc/infrastructure/utils/utils.h"
#include "analysis/csrc/infrastructure/utils/file.h"
#include "device_context.h"
#include "device_context_error_code.h"

using namespace Analysis;
using namespace Analysis::Utils;
using namespace Analysis::Domain;


namespace nlohmann {

template <>
struct adl_serializer<DeviceInfo> {
    static void to_json(json& jsonData, const DeviceInfo& infoData)
    {
        jsonData = json{{"platform_version", infoData.chipID}, {"devices", infoData.deviceId}};
    }
    static void from_json(const json& jsonData, DeviceInfo& infoData)
    {
        uint32_t fromJsonResult = static_cast<uint32_t>(ANALYSIS_OK);
        std::string platformVersionStr = jsonData.at("platform_version").get<std::string>();
        fromJsonResult |= static_cast<uint32_t>(StrToU32(infoData.chipID, platformVersionStr));

        std::string deviceStr = jsonData.at("devices").get<std::string>();
        fromJsonResult |= static_cast<uint32_t>(StrToU32(infoData.deviceId, deviceStr));

        jsonData.at("DeviceInfo").at(0).at("ai_core_num").get_to(infoData.aiCoreNum);

        std::string aicFrequencyStr = jsonData.at("DeviceInfo").at(0).at("aic_frequency").get<std::string>();
        fromJsonResult |= static_cast<uint32_t>(StrToU32(infoData.aicFrequency, aicFrequencyStr));

        jsonData.at("DeviceInfo").at(0).at("aiv_num").get_to(infoData.aivNum);

        std::string aivFrequencyStr = jsonData.at("DeviceInfo").at(0).at("aiv_frequency").get<std::string>();
        fromJsonResult |= static_cast<uint32_t>(StrToU32(infoData.aivFrequency, aivFrequencyStr));

        std::string hwtsFrequencyStr = jsonData.at("DeviceInfo").at(0).at("hwts_frequency").get<std::string>();
        fromJsonResult |= static_cast<uint32_t>(StrToDouble(infoData.hwtsFrequency, hwtsFrequencyStr));
        if (fromJsonResult != static_cast<uint32_t>(ANALYSIS_OK) || IsDoubleEqual(infoData.hwtsFrequency, 0.0)) {
            ERROR("Get device info from json error! The input: platform_version= %, devices= %, aic_frequency= %, "
                  "aiv_frequency= %, hwts_frequency=%", platformVersionStr, deviceStr, aicFrequencyStr, aivFrequencyStr,
                  hwtsFrequencyStr);
            throw std::runtime_error("Get info json failed");
        }
    }
};

}

namespace Analysis {

namespace Domain {

const std::string INFO_JSON = "info.json";

bool DeviceContext::GetInfoJson()
{
    std::vector <std::string> files = Analysis::Utils::File::GetOriginData(deviceContextInfo.deviceFilePath,
                                                                           {INFO_JSON}, {"done"});
    if (files.size() != 1) {
        ERROR("The number of % in % is invalid.", INFO_JSON, deviceContextInfo.deviceFilePath);
        return false;
    }

    FileReader fd(files.back());
    nlohmann::json info;
    if (fd.ReadJson(info) != ANALYSIS_OK) {
        ERROR("Load json context failed: '%'.", files.back());
        return false;
    }

    try {
        deviceContextInfo.deviceInfo = info;
    } catch (const std::exception& e) {
        ERROR("Error parsing JSON: '%'.", e.what());
        return false;
    }
    return true;
}

}

}

File: analysis/profiling_bean/basic_info/device_info.py
# -------------------------------------------------------------------------
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is part of the MindStudio project.
#
# MindStudio is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#
#    http://license.coscl.org.cn/MulanPSL2
#
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# -------------------------------------------------------------------------

from common_func.info_conf_reader import InfoConfReader
from profiling_bean.basic_info.base_info import BaseInfo


class DeviceInfo(BaseInfo):
    """
    device info
    """
    DEVICE_INFO = "DeviceInfo"

    # The json keys under the DeviceInfo
    KEY_DEVICE_ID = "id"
    KEY_CTRL_CPU_ID = "ctrl_cpu_id"
    KEY_CTRL_CPU_CORE_NUM = "ctrl_cpu_core_num"
    KEY_TS_CPU_CORE_NUM = "ts_cpu_core_num"
    KEY_AI_CPU_CORE_NUM = "ai_cpu_core_num"
    KEY_AI_CORE_NUM = "ai_core_num"

    def __init__(self: any) -> None:
        super(DeviceInfo, self).__init__()
        self.device_id = 0
        self.ai_cpu_num = 0
        self.ai_core_num = 0
        self.control_cpu_num = 0
        self.control_cpu_type = ""
        self.ts_cpu_num = 0

    def run(self: any, _: str) -> None:
        """
        run data
        :return: None
        """
        if not InfoConfReader().get_root_data(self.DEVICE_INFO):
            # clear all Member variables to remind the key of device_info.
            self.__dict__.clear()
        else:
            self.merge_data()

    def merge_data(self: any) -> None:
        """
        merge data
        :return:None
        """
        self.device_id = InfoConfReader().get_data_under_device(self.KEY_DEVICE_ID)
        self.control_cpu_type = InfoConfReader().get_data_under_device(
            self.KEY_CTRL_CPU_ID)
        self.control_cpu_num = InfoConfReader().get_data_under_device(
            self.KEY_CTRL_CPU_CORE_NUM)
        self.ts_cpu_num = InfoConfReader().get_data_under_device(self.KEY_TS_CPU_CORE_NUM)
        self.ai_cpu_num = InfoConfReader().get_data_under_device(self.KEY_AI_CPU_CORE_NUM)
        self.ai_core_num = InfoConfReader().get_data_under_device(self.KEY_AI_CORE_NUM)
